google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddDynamicPageFeed.java

```diff
@@ -0,0 +1,516 @@
+          "Updated campaign with resource name '%s'.%n", response.getResults(0).getResourceName());
+    }
+  }
+
+  /**
+   * Returns the DSA settings for a campaign. Throws an error if the campaign does not exist or is
+   * not a DSA campaign.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the customer ID.
+   * @param campaignId the campaign ID of the campaign to update.
+   * @return the DSA settings for the campaign.
+   */
+  private static DynamicSearchAdsSetting getDsaSetting(
+      GoogleAdsClient googleAdsClient, long customerId, long campaignId) {
+    // Creates the query.
+    // You must request all DSA fields in order to update the DSA settings in the following step.
+    String query =
+        String.format(
+            "SELECT campaign.id, campaign.name, campaign.dynamic_search_ads_setting.domain_name,"
+                + " campaign.dynamic_search_ads_setting.language_code,"
+                + " campaign.dynamic_search_ads_setting.use_supplied_urls_only "
+                + "FROM campaign "
+                + "WHERE campaign.id = '%s'",
+            campaignId);
+
+    // Creates the request.
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setCustomerId(Long.toString(customerId))
+            .setPageSize(PAGE_SIZE)
+            .setQuery(query)
+            .build();
+
+    // Creates the service client.
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // Issues the search request.
+      SearchPagedResponse response = googleAdsServiceClient.search(request);
+
+      // Throws an exception if the campaign is not a DSA campaign.
+      // The server will return an error if the campaign is not a DSA campaign, but this
+      // exception is easier to interpret.
+      if (!response
+          .getPage()
+          .getResponse()
+          .getResults(0)
+          .getCampaign()
+          .hasDynamicSearchAdsSetting()) {
+        throw new IllegalArgumentException(
+            String.format("Campaign with ID '%s' is not a " + "DSA campaign.", campaignId));
+      }
+
+      // Retrieves and returns the DSA setting.
+      return response
+          .getPage()
+          .getResponse()
+          .getResults(0)
+          .getCampaign()
+          .getDynamicSearchAdsSetting();
+    }
+  }
+
+  /**
+   * Creates an ad group criterion targeting the DSA label.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the customer ID.
+   * @param adGroupId the ad group ID.
+   * @param dsaPageUrlLabel the label for the DSA page URLs.
+   */
+  private static void addDsaTarget(
+      GoogleAdsClient googleAdsClient, long customerId, long adGroupId, String dsaPageUrlLabel) {
+    String adGroupResourceName = ResourceNames.adGroup(customerId, adGroupId);
+
+    // Creates the webpage condition info.
```

Comment:
The comments here explain what but not why. Could you make them more informative so they cover more than just the type of each object?